collector/recon.go

```diff
@@ -0,0 +1,681 @@
+	pathToReconExecutable string
+	md5OutputRx           *regexp.Regexp
+
+	all        typedDesc
+	errors     typedDesc
+	matched    typedDesc
+	notMatched typedDesc
+}
+
+func newReconMD5Task(pathToReconExecutable string) collectorTask {
+	return &reconMD5Task{
+		pathToReconExecutable: pathToReconExecutable,
+		md5OutputRx: regexp.MustCompile(
+			`(?m)^.* Checking ([\.a-zA-Z0-9_]+) md5sum(?:s)?\s*([0-9]+)/([0-9]+) hosts matched, ([0-9]+) error.*$`),
+		all: typedDesc{
+			desc: prometheus.NewDesc("swift_cluster_md5_all",
+				"Sum of matched-, not matched hosts, and errors encountered while check hosts for md5sum(s) as reported by the swift-recon tool.",
+				[]string{"kind"}, nil),
+			valueType: prometheus.GaugeValue,
+		},
+		errors: typedDesc{
+			desc: prometheus.NewDesc("swift_cluster_md5_errors",
+				"Errors encountered while checking hosts for md5sum(s) as reported by the swift-recon tool.",
+				[]string{"kind"}, nil),
+			valueType: prometheus.GaugeValue,
+		},
+		matched: typedDesc{
+			desc: prometheus.NewDesc("swift_cluster_md5_matched",
+				"Matched hosts for md5sum(s) reported by the swift-recon tool.",
+				[]string{"kind"}, nil),
+			valueType: prometheus.GaugeValue,
+		},
+		notMatched: typedDesc{
+			desc: prometheus.NewDesc("swift_cluster_md5_not_matched",
+				"Not matched hosts for md5sum(s) reported by the swift-recon tool.",
+				[]string{"kind"}, nil),
+			valueType: prometheus.GaugeValue,
+		},
+	}
+}
+
+// reconMD5Task implements the collector.collectorTask interface.
+func (t *reconMD5Task) describeMetrics(ch chan<- *prometheus.Desc) {
+	t.all.describe(ch)
+	t.errors.describe(ch)
+	t.matched.describe(ch)
+	t.notMatched.describe(ch)
+}
+
+// reconMD5Task implements the collector.collectorTask interface.
+func (t *reconMD5Task) collectMetrics(ch chan<- prometheus.Metric, exitCodeTypedDesc typedDesc) {
+	exitCode := 0
+	cmdArg := "--md5"
+	out, err := exec.Command(t.pathToReconExecutable, cmdArg).Output()
+	if err == nil {
+		matchList := t.md5OutputRx.FindAllSubmatch(out, -1)
+		if len(matchList) > 0 {
+			for _, match := range matchList {
+				var totalHosts, errsEncountered float64
+				matchedHosts, err := strconv.ParseFloat(string(match[2]), 64)
+				if err == nil {
+					totalHosts, err = strconv.ParseFloat(string(match[3]), 64)
+					if err == nil {
+						errsEncountered, err = strconv.ParseFloat(string(match[4]), 64)
+						if err == nil {
+							kind := strings.ReplaceAll(string(match[1]), ".", "")
+							notMatchedHosts := totalHosts - matchedHosts
+							all := matchedHosts + notMatchedHosts + errsEncountered
+							ch <- t.all.mustNewConstMetric(all, kind)
+							ch <- t.errors.mustNewConstMetric(errsEncountered, kind)
+							ch <- t.matched.mustNewConstMetric(matchedHosts, kind)
+							ch <- t.notMatched.mustNewConstMetric(notMatchedHosts, kind)
+						}
+					}
+				}
+				if err != nil {
+					exitCode = 1
+					logg.Error("swift dispersion: %s: %s", cmdArg, err.Error())
```

Comment:
Here, and in several other places in this file, it says `swift dispersion` in the error messages, which I assume is a copy-paste error.